core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java

```diff
@@ -202,4 +221,29 @@ public int hashCode() {
         result = 31 * result + (unit != null ? unit.hashCode() : 0);
         return result;
     }
+
+    @Override
+    public int compareTo(ByteSizeValue other) {
+        long unitValue = unit.toBytes(1);
+        long otherUnitValue = other.unit.toBytes(1);
+
+        ByteSizeUnit minUnit = unitValue < otherUnitValue ? unit : other.unit;
+        long thisValue = this.convert(minUnit);
+        long otherValue = other.convert(minUnit);
+
+        if (thisValue == otherValue) {
```

Comment:
Should I keep my original behavior and make equals defer to the new compareTo? 

(The compareTo that was implemented tries to find the minimum of the two ByteSizeUnits and converts both to them, causing this to work out properly)